knowledge-base/adr/0031-resilience-options.md

```diff
@@ -61,310 +61,153 @@ So we have something for circuit breaker and timeout.
-  timeout?: StringValue | false | undefined; // default 10sec
-  errorThresholdPercentage?: number | undefined; // default 50
-  volumeThreshold?: number | undefined; // default 10
-  resetTimeout?: number | undefined; // default 30000
-  isolationStragtegy?: IsolationStrategy; // default tenant
-}
-
-interface AsyncRetryLibOptions {
-  retries?: number; // default 10
-  factor?: number; // default  2.
-  minTimeout?: StringValue; // default 1000 ms. See https://github.com/vercel/ms
-  maxTimeout?: StringValue; // default Infinity. See https://github.com/vercel/ms
-  randomize?: boolean; // default true.
-  onRetry: (e: Error) => {}; // default undefined
-}
-
-// in the resilience call:
-type ResilienceMiddlewareOptions = {
-  timeout: (context?: RequestContext) => TimeoutOptions;
-  retry: (context?: RequestContext) => RetryOptions;
-  circuitBreaker: (context?: RequestContext) => CircuitBreakerOptions;
+type MiddlewareInOut<T> = {
+  fn: () => Promise<T>;
+  exitChain: boolean;
+  context?: RequestContext;
 };
+
+type Middleware<T> = <T>(options: MiddlewareInOut) => MiddlewareInOut;
 ```
 
-The explicit way to define the default situation with 10sec timeout, circuit breaker for BTP and retry switched off would be:
+In our code we will pass around a middleware array and wrap it around evey http call at the very end:
 
 ```ts
-defaultResilienceOptions: ResilienceMiddleWareOptions = {
-  timeout: (context: RequestContext) => 10000,
-  retry: (context: RequestContext) => false,
-  circuitBreaker: enableCircuitBreakerForBTP
-};
+async function someHttpSdkMethod(middleware: Middleware[]): Promise<any> {
+  //return someHttpCall(someArgs)
+  const context = {}; //context for this method
 
-function enableCircuitBreakerForBTP(
-  context: RequestContext
-): CircuitBreakerOptions {
-  if (context.url.includes(btpDomain)) {
-    return defaultCircuitBreakerOptions;
-  }
-  return false;
+  return joinMiddleware(middleware)({
+    context,
+    fn: someHttpCall,
+    someArgs,
+    exitChain: false
+  }).fn(someArgs);
 }
 ```
 
-The default will change in version 3.0 to:
+The method `joinMiddleware` joins all the function in an async pipe.
+The context will be a minimal but guaranteed amount of information:
 
 ```ts
-defaultResilienceOptions: ResilienceMiddleWareOptions = {
-  timeout: (context: RequestContext) => 1000,
-  retry: (context: RequestContext) => defaultRetryOptions,
-  circuitBreaker: (context: RequestContext) => defaultCircuitBreakerOptions
-};
-```
-
-The user can provide sophisticated methods if needed.
-From version 3.0 the resilience will be on by default using the options above, so the methods will be either used to disable or adjust the resilience.
-A `disabled` method for convenience could be nice:
-
-```ts
-function disabled(context: RequestContext): false {
-  return false;
+export interface Context {
+  category: 'xsuaa' | 'destination' | 'user-defined';
+  destination: Destination; //only url is mandatory
+  tenantId: string;
```

Comment:
here I do not understand what you mean?

Reply:
It looks to me that the user needs to specify a context for the middlewares. So, are they also responsible for providing a `tenantId` as part of the context? If so, they can pick any `tenantId` as they want and this might be unsafe. Instead of letting `tenantId` be the property of the context, we (the SDK) can specify the `tenantId` from the `jwt` automatically.

Reply:
The middle ware code should be part of the actual implementation. So if people do something there it is not our issue. You should not allow any kind of generic code injection in the middle ware :-).